src/big/big_decimal.cr

```diff
@@ -0,0 +1,287 @@
+    initialize num.to_s # to appease the compiler
+  end
+
+  # Creates a new `BigDecimal` from `BigInt`/`UInt64`, which matches the internal representation
+  def initialize(@value : BigInt, @scale : UInt64)
+  end
+
+  def initialize(value : Int, scale : Int)
+    initialize(value.to_big_i, scale.to_u64)
+  end
+
+  def initialize(value : BigInt)
+    initialize(value, 0u64)
+  end
+
+  def initialize
+    initialize(0)
+  end
+
+  # Returns *num*. Useful for generic code that does `T.new(...)` with `T`
+  # being a `Number`.
+  def self.new(num : BigDecimal)
+    num
+  end
+
+  def +(other : BigDecimal) : BigDecimal
+    if @scale > other.scale
+      scaled = other.scale_to(self)
+      BigDecimal.new(@value + scaled.value, @scale)
+    elsif @scale < other.scale
+      scaled = scale_to(other)
+      BigDecimal.new(scaled.value + other.value, other.scale)
+    else
+      BigDecimal.new(@value + other.value, @scale)
+    end
+  end
+
+  def -(other : BigDecimal) : BigDecimal
+    if @scale > other.scale
+      scaled = other.scale_to(self)
+      BigDecimal.new(@value - scaled.value, @scale)
+    elsif @scale < other.scale
+      scaled = scale_to(other)
+      BigDecimal.new(scaled.value - other.value, other.scale)
+    else
+      BigDecimal.new(@value - other.value, @scale)
+    end
+  end
+
+  def *(other : BigDecimal) : BigDecimal
+    BigDecimal.new(@value * other.value, @scale + other.scale)
+  end
+
+  def /(other : BigDecimal) : BigDecimal
+    div other
+  end
+
+  # Divides self with another `BigDecimal`, with a optionally configurable max_div_iterations, which
+  # defines a maximum number of iterations in case the division is not exact.
+  #
+  # ```
+  # BigDecimal(1).div(BigDecimal(2)) => BigDecimal(@value=5, @scale=2)
+  # BigDecimal(1).div(BigDecimal(3), 5) => BigDecimal(@value=33333, @scale=5)
+  # ```
+  def div(other : BigDecimal, max_div_iterations = DEFAULT_MAX_DIV_ITERATIONS) : BigDecimal
+    check_division_by_zero other
+
+    scale = @scale - other.scale
+    numerator, denominator = @value, other.@value
+
+    quotient, remainder = numerator.divmod(denominator)
+    if remainder == ZERO
+      return BigDecimal.new(normalize_quotient(other, quotient), scale)
+    end
+
+    remainder = remainder * TEN
+
+    i = 0
+    while remainder != ZERO && i < max_div_iterations
+      inner_quotient, inner_remainder = remainder.divmod(denominator)
+      quotient = quotient * TEN + inner_quotient
+      remainder = inner_remainder * TEN
+      i += 1
+    end
+
+    BigDecimal.new(normalize_quotient(other, quotient), scale + i)
+  end
+
+  def <=>(other : BigDecimal) : Int32
+    if @scale > other.scale
+      @value <=> other.scale_to(self).value
+    elsif @scale < other.scale
+      scale_to(other).value <=> other.value
+    else
+      @value <=> other.value
+    end
+  end
+
+  def <=>(other : Int)
+    @value <=> other
```

Comment:
I'm confused, isn't this incorrect? Wouldn't `@value` for `6.1234` be `61234`, and so if you compare it to `7_u32` you're comparing 61234 with 7, so it'll say `6.1234 > 7` ? I don't see any specs which do `BigDecimal.new(6.5) < 7` for example, they're all comparing bigdecimal ints with ints.

Reply:
Oh, indeed it is. I added that unit test and corrected the implementation as `self <=> BigDecimal.new(other)`